### DISCUSSION TITLE
null is not an object (evaluating 'b.parentNode') in production

### DISCUSSION THREAD

**Discussion by Fonger** [vercel/next.js#70048] [App Router]
I don't know how to reproduce but it seems it only happen in Safari.

My Sentry keeps receiving reports from users like this

```
app:///xxxx(path-to-the-page) in $RS at line 1:119385
```

with error message:
```
null is not an object (evaluating 'b.parentNode')
```

I search the HTML document reporting the error, I see

```html
<script>$RS=function(a,b){a=document.getElementById(a);b=document.getElementById(b);for(a.parentNode.removeChild(a);a.firstChild;)b.parentNode.insertBefore(a.firstChild,b);b.parentNode.removeChild(b)};
```

This indicates that param b, which is `document.getElementById(b)` is somehow null.

This code seems to be replated to streaming rendering and this is how the server-sent data push to the client suspense `<template id="xxx"/>` slot.

Is it possible? how come `$RS()` calling tries to to insert a non-existing template slot id?

I can't produce it but sentry logs indicate the following environments. All of them are from Sarari.

Note these data with 3% sample rate and this indicating that the actual affected users are 33 times larger than this. And it's within 24hr.

![image](https://github.com/user-attachments/assets/3e9a456d-5dd1-48c3-846f-69d0ed4ec859)

**Comment by Behnoudmst** [comment]
I am having same issue on Nextjs 14.2

**Comment by icyJoseph** [comment]
This seems to come from React itself... 🤔 Could it be a browser extension, or perhaps people who tell their browser to translate the site? 

- https://github.com/vercel/next.js/blob/09aa303a76f4441b497444ba25db024216e771ea/packages/next/src/compiled/react-dom/cjs/react-dom-server.node.production.js#L2277-L2282

**Reply by icyJoseph** [reply]
@Fonger @Behnoudmst I went over to the React repo, and found this, https://github.com/facebook/react/issues/25710

Apparently, wrapping the entire HTML document with a Suspense boundary, could cause this issue: https://github.com/facebook/react/issues/25710#issuecomment-1352532140

**Comment by vitalyiegorov** [comment]
facing same issue with the `Next 15.3.1`

**Reply by serhii-kucherenko** [reply]
Same with the `15.1.6`

**Reply by leog** [reply]
Same with `15.5.2`

**Reply by icyJoseph** [reply]
Any repositories? Pseucode to look at? Do we know what Safari versions trigger this?

**Reply by leog** [reply]
Not really, sorry @icyJoseph. And in my case is not just Safari.
Could it be related to having a client component for the loading.tsx file?
As the loading animation on our case depends on the accessed path, and the loading.tsx doesn't get dynamic routes values, I needed to make it a client component.

**Reply by icyJoseph** [reply]
Is there a publicly available site? Could try to debug there if so. Worth trying latest 15.5.6, on the way there React got some updates, which might have fixed this.

**Reply by leog** [reply]
Created a PR for my app which has a PR preview upgrading to Next.js 15.5.6 and still happening. Managed to move things around to rely on server side `loading.tsx` file and still happening. The app continues to work well, but Sentry is going crazy with this error. 

Maybe there is a way the line mentioned above and the other flavors (bun, etc) like the following https://github.com/vercel/next.js/blob/bc59f210b02256e63b93d9f473c97ae130568e63/packages/next/src/compiled/react-dom/cjs/react-dom-server.bun.production.js#L6523 can be patched to not throw an error when there is no elements in `a` or `b` which results in a null reference. Seems like an oversight to me, even if it is not expected to have an empty element, it's safer to assume it could not have an element and not apply the change instead of erroring out. I could create a PR to patch those files but doesn't seem right; those are part of the compiled version of `next.js` under the `react-dom` dependency. 

Found this other instance of the same issue https://github.com/facebook/react/issues/25710 by @tannerlinsley.

cc: @timneutkens who has helped me with next.js issues in the past.

**Reply by icyJoseph** [reply]
@leog That issue from the React repository, was also mentioned here, https://github.com/vercel/next.js/discussions/70048#discussioncomment-11519849 and it looks like it comes from placing a Suspense outside the body?

>  It's specifically <Suspense> node outside <body> that's not supported

**Reply by leog** [reply]
@icyJoseph sorry for the delay, and it's not my case, don't have any Suspense outside the body

**Reply by leog** [reply]
@icyJoseph found the spot to fix it, here it is https://github.com/facebook/react/pull/34996 in case there's a chance Vercel/Next can pull some strings to rush things a bit

**Comment by leog** [comment]
Fix is here: https://github.com/facebook/react/pull/34996

**Comment by RaghuSai-09** [comment]
Root cause found & Fix available (Update to 15.5.6+)

Hey everyone, I’ve been digging into this, and it looks like a classic hydration/streaming mismatch.

The error happens because the $RS (React Streaming) function is trying to swap out a Suspense boundary placeholder, but it can't find the target element in the DOM yet—this is why we're seeing b.parentNode return null. It seems especially common in Safari or when browser extensions (like Google Translate) mess with the DOM structure during the stream.

Here is how to clear this up:

The Official Fix: This was recently addressed in the React core. If you update to Next.js 15.5.6 or later, you'll get the synchronized React changes that include a null check in the $RS function to prevent this crash.

Check your Layout: Make sure you don't have any <Suspense> boundaries wrapping your <html> or <body> tags in layout.tsx. As mentioned earlier in the thread, putting nodes outside the body isn't supported and is a major trigger for this.

Third-Party Scripts: If you only see this in prod, check if you have any analytics or "accessibility" scripts that modify the DOM immediately on load. They might be "stealing" the element before React can finish the swap.

Technical Reference:
The specific fix is in [React PR #34996](https://github.com/facebook/react/pull/34996).

Hope this helps clear out those Sentry logs!
